File: 程式設計：資料結構答案檔/Go/103.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Monster struct {
	Value int
	Next  *Monster
}

type BattleSimulator struct {
	head *Monster
}

func NewBattleSimulator() *BattleSimulator {
	return &BattleSimulator{}
}

func (bs *BattleSimulator) CreateMonsters(input string) {
	values := strings.Split(input, ",")
	for _, value := range values {
		if monsterAttack, err := strconv.Atoi(value); err == nil {
			bs.AddMonster(monsterAttack)
		} else {
			fmt.Println("input error")
		}
	}
}

func (bs *BattleSimulator) AddMonster(value int) {
	newNode := &Monster{Value: value}
	if bs.head == nil {
		bs.head = newNode
	} else {
		current := bs.head
		for current.Next != nil {
			current = current.Next
		}
		current.Next = newNode
	}
}

func (bs *BattleSimulator) SimulateBattle(playerHP int) bool {
	current := bs.head
	level := 1

	for current != nil {
		playerHP -= current.Value

		if playerHP <= 0 {
			fmt.Printf("game over! dead at level %d\n", level)
			return false
		}

		level++
		current = current.Next
	}

	fmt.Printf("game clear! hp left %d\n", playerHP)
	return true
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	if scanner.Scan() {
		input := scanner.Text()

		battle := NewBattleSimulator()
		battle.CreateMonsters(input)

		playerHP := 100
		battle.SimulateBattle(playerHP)
	}
}

File: 程式設計：資料結構答案檔/Python/103.py
class Monster:
    def __init__(self, value):
        self.value = value
        self.next = None

class BattleSimulator:
    def __init__(self):
        self.head = None

    def create_monsters(self, input_str):
        values = input_str.split(',')
        for value in values:
            try:
                monster_attack = int(value)
                self.add_monster(monster_attack)
            except ValueError:
                print("input error")

    def add_monster(self, value):
        new_node = Monster(value)
        if not self.head:
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node

    def simulate_battle(self, player_hp):
        current = self.head
        level = 1
        while current:
            player_hp -= current.value
            if player_hp <= 0:
                print(f"game over! dead at level {level}")
                return False
            level += 1
            current = current.next
        print(f"game clear! hp left {player_hp}")
        return True

if __name__ == "__main__":
    input_str = input()
    battle = BattleSimulator()
    battle.create_monsters(input_str)
    player_hp = 100
    battle.simulate_battle(player_hp)
